### PR TITLE
[flang] Diagnose the impure procedure reference in finalization according to the rank of the entity

### PR DESCRIPTION
**Pull request by kkwli**
Use the rank of the array section to determine which final procedure would be called in diagnosing whether that procedure is impure or not.

### CODE REVIEW

File: flang/lib/Semantics/check-do-forall.cpp [hunk 1]
```diff
@@ -220,8 +220,11 @@ class DoConcurrentBodyEnforce {
       if (MightDeallocatePolymorphic(*entity, DeallocateNonCoarray)) {
         SayDeallocateOfPolymorph(variable.GetSource(), *entity, reason);
       }
-      if (const Symbol * impure{HasImpureFinal(*entity)}) {
-        SayDeallocateWithImpureFinal(*entity, reason, *impure);
+      if (const auto *assignment{GetAssignment(stmt)}) {
+        const auto lhs{assignment->lhs};
```
Comment (klausler) [inline flang/lib/Semantics/check-do-forall.cpp:224]: This should probably be a const reference `const auto &lhs{...}`.

File: flang/lib/Semantics/tools.cpp [hunk 2]
```diff
@@ -827,15 +827,21 @@ static const Symbol *HasImpureFinal(
...
       if (const DerivedTypeSpec * derived{symType->AsDerived()}) {
-        // finalizable assumed-rank not allowed (C839)
-        return evaluate::IsAssumedRank(symbol)
-            ? nullptr
-            : HasImpureFinal(*derived, symbol.Rank());
+        if (evaluate::IsAssumedRank(symbol)) {
+          // finalizable assumed-rank not allowed (C839)
+          return nullptr;
+        } else {
+          int actualRank = symbol.Rank();
```
Comment (klausler) [inline flang/lib/Semantics/tools.cpp:839]: `int actualRank{rank.value_or(symbol.Rank())};`

**Review by klausler — CHANGES_REQUESTED** [review]

File: flang/lib/Semantics/check-do-forall.cpp (hunk 1 above)
Comment (kkwli) [reply flang/lib/Semantics/check-do-forall.cpp]: Changed.

File: flang/lib/Semantics/tools.cpp (hunk 2 above)
Comment (kkwli) [reply flang/lib/Semantics/tools.cpp]: Updated

File: flang/lib/Semantics/tools.cpp [hunk 3]
```diff
@@ -827,15 +827,21 @@ static const Symbol *HasImpureFinal(
...
       if (const DerivedTypeSpec * derived{symType->AsDerived()}) {
-        // finalizable assumed-rank not allowed (C839)
-        return evaluate::IsAssumedRank(symbol)
-            ? nullptr
-            : HasImpureFinal(*derived, symbol.Rank());
+        if (evaluate::IsAssumedRank(symbol)) {
+          // finalizable assumed-rank not allowed (C839)
+          return nullptr;
+        } else {
+          int actualRank{symbol.Rank()};
```
Comment (klausler) [inline flang/lib/Semantics/tools.cpp:839]: `int actualRank{rank.value_or(symbol.Rank())};`

**Review by klausler — CHANGES_REQUESTED** [review]

File: flang/lib/Semantics/tools.cpp (hunk 3 above)
Comment (kkwli) [reply flang/lib/Semantics/tools.cpp]: Oh! Got it.

**Review by klausler — APPROVED** [review]

**Comment by kkwli** [comment]
Thanks for the review.

**Pull request merged**